### PR TITLE
Update cart currency when order currency changes

### LINKED ISSUE #18433: BO - Wrong total in order detail page when adding a product with modified price
#### Describe the bug

When adding a product with modified price to an order in BO, the "Total" price is wrong : It's the catalog price which is applied, instead of the modified price.

The "Products" total is OK.

#### Expected behavior

The Total should be with the modified price.

#### Steps to Reproduce

1. Go to FO, create an order.
2. Go to BO, to the created order, click on "Add a product" button
3. Select a product and change it's price
4. Add the product : The "Total" is wrong, when "Products" is right.

**Screenshots**

https://drive.google.com/file/d/1gH9Y2xsDCphQXkUmpKOIXBWYjjPIC0x4/view?usp=sharing

#### Additional information

* PrestaShop version: 1.7.7
* PHP version: N/A

### LINKED ISSUE #18443: BO - Wrong total when changing currency after placing the order and adding a product
#### Describe the bug

When adding a BO product to an order for which the currency has been changed, the order total is displayed in the original currency instead of the new currency.

This is an issue linked to this PR : https://github.com/PrestaShop/PrestaShop/pull/17927. I missed it when I tested the PR, sorry for that.

#### Expected behavior

The total should be displayed in the new currency.

#### Steps to Reproduce

1. Go to BO, add a second currency to the shop
2. Go to FO, make an order with any currency
3. Go to BO, to the order detail
4. In Payment block, change the currency. The change is OK, all Total and prices are converted
5. Add a product to the order : The totals are wrong : The sign is OK, but the number are in the first currency instead of the new one.

**Screenshots**

https://drive.google.com/file/d/16AtymmvvwpaNCO06UBdUmtOTKYFBwJQz/view?usp=sharing

#### Additional information

* PrestaShop version: 1.7.7
* PHP version: N/A

### PR DESCRIPTION
**Pull request by sowbiba**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | When order currency is changed, the linked cart's currency remains the old one and creates prices issues when a product is added to the order
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #18433 and Fixes #18443 
| How to test?  | #18443<br>- Go to BO, add a second currency to the shop<br>- Go to FO, make an order with any currency<br>- Go to BO, to the order detail<br>- In Payment block, change the currency. The change is OK, all Total and prices are converted<br>- Add a product to the order : The totals are wrong : The sign is OK, but the number are in the first currency instead of the new one.<br><br>#18433<br>- First create a secondary currency for your shop<br>- Go to FO, create an order with the secondary currency.<br>- Go to BO, to the created order, click on "Add a product" button<br>- Select a product and change it's price<br>- Add the product : The "Total" is wrong, when "Products" is right.<br>- The behaviour must be correct for the HT and TTC price modifications

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/18456)

### CODE REVIEW

**Comment by matks** [comment]
@sowbiba Code looks good, but I have a question.

If I understand correctly, even after cart has been placed and transformed into an order, the cart remains and is linked to the order.
When we add products to the order, the code finds the cart, puts the products into the cart (and then, I dont know how, it will pass this data to the order).

My question is : this cart, is it now "fully linked" to this order only and unreachable from anywhere else ?
- no other orders can be linked to this cart ?
- from its front-office "my account" area, customer cannot see it anymore ?
- from its front-office "my account" area, customer cannot modify it anymore ?

**Comment by sowbiba** [comment]
> @sowbiba Code looks good, but I have a question.
> 
> If I understand correctly, even after cart has been placed and transformed into an order, the cart remains and is linked to the order.
> When we add products to the order, the code finds the cart, puts the products into the cart (and then, I dont know how, it will pass this data to the order).
> 
> My question is : this cart, is it now "fully linked" to this order only and unreachable from anywhere else ?
> 
> * no other orders can be linked to this cart ?
> * from its front-office "my account" area, customer cannot see it anymore ?
> * from its front-office "my account" area, customer cannot modify it anymore ?

When a product is added to an existing order, it's the cart which is used to compute prices

https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php#L140

and the cart currency

https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Cart/CartRow.php#L297

**Comment by sowbiba** [comment]
> @sowbiba Code looks good, but I have a question.
> 
> If I understand correctly, even after cart has been placed and transformed into an order, the cart remains and is linked to the order.
> When we add products to the order, the code finds the cart, puts the products into the cart (and then, I dont know how, it will pass this data to the order).
> 
> My question is : this cart, is it now "fully linked" to this order only and unreachable from anywhere else ?
> 
> * no other orders can be linked to this cart ?
> * from its front-office "my account" area, customer cannot see it anymore ?
> * from its front-office "my account" area, customer cannot modify it anymore ?

You are right about "no other orders can be linked to this cart ?"

I didn't think about that case. When the currency is changed it's only for the current order

To avoid this, we can pass the currency to use to the calculator (the order currency)

https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Cart.php#L1898

https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Cart.php#L1968

I need to measure the impact of this

File: src/Adapter/Order/CommandHandler/ChangeOrderCurrencyHandler.php [hunk 1]
```diff
@@ -171,6 +173,19 @@ private function updateOrder(Order $order, Currency $oldCurrency, Currency $newC
         $order->update();
     }
 
+    /**
+     * @param int $cartId
+     * @param Currency $oldCurrency
+     * @param Currency $newCurrency
+     */
+    private function updateCart(int $cartId, Currency $oldCurrency, Currency $newCurrency): void
+    {
```
Comment (matthieu-rolland) [inline src/Adapter/Order/CommandHandler/ChangeOrderCurrencyHandler.php:181]: It seems like you don't need the $oldCurrency parameter after all ;)

**Review by jolelievre — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Comment by sarahdib** [comment]
@sowbiba 

hum the bug is still there 
Capture d’écran 2020-04-06 à 15 15 18

**Comment by sowbiba** [comment]
> @sowbiba
> 
> hum the bug is still there
> Capture d’écran 2020-04-06 à 15 15 18

:thinking: 
I did 2 tests : from default currency to secondary one and from secondary to default
Seems good to me. Maybe we dont test it the same way or I miss something in the scenario

https://drive.google.com/file/d/15jY-0Z3OWEX0NzEhnA9jfQZx6DcdZapE/view

https://drive.google.com/file/d/1kVokrYpK5IfQnLAFNtO-AJs4C3zwhmNY/view

**Review by Progi1984 — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Comment by sarahdib** [comment]
- Go to BO, add dollar currency to the shop
- Create a cart rule (-10€)
- Go to FO, make an order with euro
- Go to BO, to the order detail
- In Payment block, change the currency to dollar. Change are ok
- Add a product to the order 
- See the cart rule isn't apply anymore And the display of the cart rule is wrong in the listing it's 10$ and in the total is 10,93$
Capture d’écran 2020-04-10 à 11 11 30


- Go to BO, add dollar currency to the shop
- Go to FO, make an order with euro
- Go to BO, to the order detail
- In Payment block, change the currency to dollar. Change are ok
- Add a product to the order 
- In Payment block, change the currency to euro.  
- See the round problem (48,74+8,40= 57,14 not 57,15)
Capture d’écran 2020-04-10 à 11 26 30

- First create a secondary currency for your shop
- Go to FO, create an order with the secondary currency.
- Go to BO, to the created order, click on "Add a product" button
- Select a product and change it's price
- See the total is wrong
Capture d’écran 2020-04-10 à 11 34 41

- First create a secondary currency for your shop
- Go to FO, create an order with the secondary currency.
- Go to BO, to the created order
- See the rounding error  (31.21+9.18-5.46=34.93 and not 34.92)
Capture d’écran 2020-04-10 à 11 37 27

**Review by matks — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by sowbiba** [comment]
@sarahdib The rounding issues will be handled by https://github.com/PrestaShop/PrestaShop/pull/18617

**Review by PierreRambaud — DISMISSED** [review]

**Comment by eternoendless** [comment]
This PR needs to be rebased

**Review by Progi1984 — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by sarahdib** [comment]
@sowbiba 

- Go to BO, add a second currency to the shop
- Create a cart rule (-10€)
- Go to FO, make an order with euro
- Go to BO, to the order detail
- In Payment block, change the currency
- check the value of the cart rule is not correct
https://drive.google.com/file/d/1md2BXTaVlYOlKWeb6QoFAaIylVxrT9B_/view

This happen only if you change the currency in the order. If you make an order using the second currency the cart rule is OK 

----------------------------------------------------------------------
- Go to FO, make an order 
- Go to BO, to the order detail
- add a product without changing the price
- add a second product and change the price
- the total is not correct
https://drive.google.com/file/d/1u6i9h8xhvjsoqm3dY4NwmEnrtaTWuYnj/view

This happen with all currencies, with a cart rule, with shipping free or not.

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by Progi1984** [comment]
@sowbiba Your PR broke a behat test : https://travis-ci.com/github/PrestaShop/PrestaShop/jobs/324017260#L4849

File: classes/SpecificPrice.php [hunk 2]
```diff
@@ -750,13 +750,7 @@ public function duplicate($id_product = false)
...
     public static function isFeatureActive()
     {
-        static $feature_active = null;
-
-        if ($feature_active === null) {
-            $feature_active = Configuration::get('PS_SPECIFIC_PRICE_FEATURE_ACTIVE');
-        }
-
-        return $feature_active;
+        return Configuration::get('PS_SPECIFIC_PRICE_FEATURE_ACTIVE');
```
Comment (PierreRambaud) [inline classes/SpecificPrice.php]:
```suggestion
        return (bool) Configuration::get('PS_SPECIFIC_PRICE_FEATURE_ACTIVE');
```

Comment (PierreRambaud) [reply classes/SpecificPrice.php]: I love removing this kind of static cache :heart:

Comment (sowbiba) [reply classes/SpecificPrice.php]: and it introduiced a bug ^^

**Review by matthieu-rolland — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Comment by Progi1984** [comment]
Thanks @sowbiba

**Pull request merged**